### PR TITLE
[BEAM-9547] Auto-populate any unimplemented methods/properties with not_implemented_method

### PR DESCRIPTION
**Pull request by TheNeuralBit**
This PR primarily adds an `@populate_not_implemented(pandas_type)` annotation for use on our Deferred pandas types. It automatically populates any missing methods and properties with `not_implemented_method`. This has a couple of benefits:
- It's more helpful for our users, directing them to BEAM-9547
- Reduces the number of failing tests, getting us closer to running this continuously
- It gives us a frequency count for unimplemented methods in the user guide test:

```
4619 total test cases:
  0 skipped (0.0%)
  784 NotImplementedError (17.0%)
    223 NameError following NotImplementedError (28.4%)
    26 'resample' is not yet supported (BEAM-9547) (3.3%)
    24 'rolling' is not yet supported (BEAM-9547) (3.1%)
    22 'cat' is not yet supported (BEAM-9547) (2.8%)
    22 'apply' is not yet supported (BEAM-9547) (2.8%)
    21 'loc' is not yet supported (BEAM-9547) (2.7%)
    20 'reindex' is not yet supported (BEAM-9547) (2.6%)
    19 'dt' is not yet supported (BEAM-9547) (2.4%)
    17 'interpolate' is not yet supported (BEAM-9547) (2.2%)
    16 'sort_index' is not yet supported (BEAM-9547) (2.0%)
    16 'loc.setitem' is not yet supported (BEAM-9547) (2.0%)
    15 'iloc' is not yet supported (BEAM-9547) (1.9%)
    13 'mean' is not yet supported (BEAM-9547) (1.7%)
    13 '__getitem__' is not yet supported (BEAM-9547) (1.7%)
    12 GroupBy.agg currently only supports callable arguments (1.5%)
    11 'ndim' is not yet supported (BEAM-9547) (1.4%)
    11 'describe' is not yet supported (BEAM-9547) (1.4%)
    11 'sample' is not yet supported (BEAM-9547) (1.4%)
    9 <class 'str'> (1.1%)
    9 'duplicated' is not yet supported (BEAM-9547) (1.1%)

<omitted for brevity>

  432 apache_beam.dataframe.frame_base.WontImplementError (9.4%)
    190 NameError following apache_beam.dataframe.frame_base.WontImplementError (44.0%)
    65 order-sensitive (15.0%)
    35 non-deferred (8.1%)
    30 non-deferred value (6.9%)
    26 order sensitive (6.0%)
    22 Conversion to a non-deferred a numpy array. (5.1%)
    17 non-lazy (3.9%)
    16 Use iloc or loc with integer slices. (3.7%)
    13 iloc.setitem (3.0%)
    8 non-deferred column values (1.9%)
    4 deprecated (0.9%)
    3 random access files (0.7%)
    1 scalar value (0.2%)
    1 plot (0.2%)
    1 others must be None, Series, or List[Series]. List[str] is not supported. (0.2%)
  1011 failed (21.9%)
  2392 passed (51.8%)
```

The PR also includes a change to doctests.py to correctly distinguish between NotImplemented and WontImplement when both are allowed, as well as some minor compliance changes, in separate commits.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by TheNeuralBit** [comment]
R: @robertwb

File: sdks/python/apache_beam/dataframe/frames.py [hunk 1]
```diff
@@ -1391,30 +1397,22 @@ def _is_associative(func):
...
-  dtypes = frame_base.not_implemented_method('dtypes')
-  expanding = frame_base.not_implemented_method('expanding')
-  ffill = frame_base.not_implemented_method('ffill')
   fillna = frame_base._elementwise_method('fillna')
   filter = frame_base._elementwise_method('filter')
   first = frame_base.wont_implement_method('order sensitive')
   get_group = frame_base._elementwise_method('group')
   head = frame_base.wont_implement_method('order sensitive')
-  hist = frame_base.wont_implement_method('plot')
+  hist = frame_base.wont_implement_method('hist')
```
Comment (robertwb) [inline sdks/python/apache_beam/dataframe/frames.py:1415]: This was intentionally 'plot' for better summarization (e.g. to see that x% of methods that we won't implement are due to them being plotting). A traceback would point to the actual method used. But if we want to be more specific, that's OK too.

File: sdks/python/apache_beam/dataframe/doctests.py [hunk 2]
```diff
@@ -253,27 +251,37 @@ def sort_and_normalize(text):
...
-    elif got.startswith('NameError') and self._seen_error:
-      # After raising WontImplementError or NotImplementError,
-      # ignore a NameError.
-      # This allows us to gracefully skip tests like
-      #    >>> res = df.unsupported_operation()
-      #    >>> check(res)
-      return True
-    else:
-      self.reset()
+
```
Comment (robertwb) [inline sdks/python/apache_beam/dataframe/doctests.py:259]: Nice cleanup, this is much better.

**Review by robertwb — APPROVED** [review]
Thanks! Makes a lot of sense to auto-populate not implemented.

**Comment by TheNeuralBit** [comment]
Run Python PreCommit

**Pull request merged**